paddle/fluid/operators/inplace_abn_op.cc

```diff
@@ -0,0 +1,208 @@
+    const auto* var = ctx.InputVar(framework::GradVarName("Y"));
+    auto input_data_type = ctx.Input<Tensor>("Y")->type();
+    if (var == nullptr) {
+      PADDLE_THROW(platform::errors::InvalidArgument(
+          "can't find gradient variable of Y"));
+    }
+    const Tensor* t = nullptr;
+    if (var->IsType<Tensor>()) {
+      t = &var->Get<Tensor>();
+    } else if (var->IsType<LoDTensor>()) {
+      t = &var->Get<LoDTensor>();
+    }
+    if (t == nullptr) {
+      PADDLE_THROW(
+          platform::errors::InvalidArgument("gradient variable of Y is empty"));
+    }
+    framework::LibraryType library = framework::LibraryType::kPlain;
+    framework::DataLayout layout = framework::DataLayout::kAnyLayout;
+
+    return framework::OpKernelType(input_data_type, ctx.GetPlace(), layout,
+                                   library);
+  }
+};
+
+class InplaceABNOpMaker : public paddle::operators::BatchNormOpMaker {
+ public:
+  void Make() override {
+    BatchNormOpMaker::Make();
+    AddAttr<std::string>(
+        "activation",
+        "(enum string, default identity, can be identity|elu|leakyrelu) "
+        "The activation type used for output candidate {h}_t.")
+        .SetDefault("");
+    AddAttr<float>("alpha",
+                   "(float, default 1.0) Only used in inplace-abn kernel,"
+                   "the activation type(identity|elu|leakyrelu) would be fused "
+                   "with batch_norm, "
+                   "this is the alpha value for elu|leakyrelu.")
+        .SetDefault(0.1f);
+    AddAttr<bool>("use_sync_bn",
+                  "(bool, default false) Whether use synchronize batch "
+                  "normalization.")
+        .SetDefault(false);
+  }
+};
+
+template <typename T>
+class InplaceABNOpGradMaker : public framework::SingleGradOpMaker<T> {
+ public:
+  using framework::SingleGradOpMaker<T>::SingleGradOpMaker;
+
+ protected:
+  void Apply(GradOpPtr<T> op) const override {
+    op->SetType(this->ForwardOpType() + "_grad");
+    op->SetInput("Y", this->Output("Y"));
+    op->SetInput(framework::GradVarName("Y"), this->OutputGrad("Y"));
+
+    op->SetInput("Scale", this->Input("Scale"));
+    op->SetInput("Bias", this->Input("Bias"));
+    op->SetInput("SavedMean", this->Output("SavedMean"));
+    op->SetInput("SavedVariance", this->Output("SavedVariance"));
+
+    // used when setting use_global_stats True during training
+    if (boost::get<bool>(this->GetAttr("use_global_stats"))) {
+      op->SetInput("Mean", this->Output("MeanOut"));
+      op->SetInput("Variance", this->Output("VarianceOut"));
+    }
+
+    op->SetAttrMap(this->Attrs());
+
+    op->SetOutput(framework::GradVarName("X"), this->InputGrad("X"));
+    op->SetOutput(framework::GradVarName("Scale"), this->InputGrad("Scale"));
+    op->SetOutput(framework::GradVarName("Bias"), this->InputGrad("Bias"));
+  }
+};
+
+DECLARE_INPLACE_OP_INFERER(InplaceABNInplaceInferer, {"X", "Y"});
+DECLARE_INPLACE_OP_INFERER(InplaceABNGradInplaceInferer,
```

Comment:
If you add `PADDLE_ENFORCE` inside kernel, please remove this unnecessary configurations of InplaceOpInferer.

Reply:
Done, thanks!